plugins/obs-webrtc/whip-output.cpp

```diff
@@ -0,0 +1,487 @@
+		peer_connection = -1;
+		audio_track = -1;
+		video_track = -1;
+		return;
+	}
+
+	obs_output_begin_data_capture2(output);
+	running = true;
+}
+
+void WHIPOutput::SendDelete()
+{
+	if (resource_url.empty()) {
+		do_log(LOG_DEBUG,
+		       "No resource URL available, not sending DELETE");
+		return;
+	}
+
+	struct curl_slist *headers = NULL;
+	if (!bearer_token.empty()) {
+		auto bearer_token_header =
+			std::string("Authorization: Bearer ") + bearer_token;
+		headers =
+			curl_slist_append(headers, bearer_token_header.c_str());
+	}
+
+	CURL *c = curl_easy_init();
+	curl_easy_setopt(c, CURLOPT_HTTPHEADER, headers);
+	curl_easy_setopt(c, CURLOPT_URL, resource_url.c_str());
+	curl_easy_setopt(c, CURLOPT_CUSTOMREQUEST, "DELETE");
+	curl_easy_setopt(c, CURLOPT_TIMEOUT, 8L);
+
+	auto cleanup = [&]() {
+		curl_easy_cleanup(c);
+		curl_slist_free_all(headers);
+	};
+
+	CURLcode res = curl_easy_perform(c);
+	if (res != CURLE_OK) {
+		do_log(LOG_WARNING,
+		       "DELETE request for resource URL failed. Reason: %s",
+		       curl_easy_strerror(res));
+		cleanup();
+		return;
+	}
+
+	long response_code;
+	curl_easy_getinfo(c, CURLINFO_RESPONSE_CODE, &response_code);
+	if (response_code != 200) {
+		do_log(LOG_WARNING,
+		       "DELETE request for resource URL failed. HTTP Code: %ld",
+		       response_code);
+		cleanup();
+		return;
+	}
+
+	do_log(LOG_DEBUG,
+	       "Successfully performed DELETE request for resource URL");
+	resource_url.clear();
+	cleanup();
+}
+
+void WHIPOutput::StopThread(bool signal)
+{
+	if (peer_connection != -1) {
+		rtcDeletePeerConnection(peer_connection);
+		peer_connection = -1;
+		audio_track = -1;
+		video_track = -1;
+	}
+
+	SendDelete();
+
+	// "signal" exists because we have to preserve the "running" state
+	// across reconnect attempts. If we don't emit a signal if
+	// something calls obs_output_stop() and it's reconnecting, you'll
+	// desync the UI, as the output will be "stopped" and not
+	// "reconnecting", but the "stop" signal will have never been
+	// emitted.
+	if (running && signal) {
+		obs_output_signal_stop(output, OBS_OUTPUT_SUCCESS);
+		running = false;
+	}
+
+	total_bytes_sent = 0;
+	connect_time_ms = 0;
+	start_time_ns = 0;
+	last_audio_timestamp = 0;
+	last_video_timestamp = 0;
+}
+
+void WHIPOutput::Send(void *data, uintptr_t size, uint64_t duration, int track)
+{
+	if (!running)
+		return;
+
+	// sample time is in us, we need to convert it to seconds
+	auto elapsed_seconds = double(duration) / (1000.0 * 1000.0);
+
+	// get elapsed time in clock rate
+	uint32_t elapsed_timestamp = 0;
+	rtcTransformSecondsToTimestamp(track, elapsed_seconds,
+				       &elapsed_timestamp);
+
+	// set new timestamp
+	uint32_t current_timestamp = 0;
+	rtcGetCurrentTrackTimestamp(track, &current_timestamp);
+	rtcSetTrackRtpTimestamp(track, current_timestamp + elapsed_timestamp);
+
+	total_bytes_sent += size;
+	rtcSendMessage(track, reinterpret_cast<const char *>(data), (int)size);
```

Comment:
@Sean-Der Maybe I'm missing something, but the track might not actually be open here. If the peer connection is still connecting, it will silently fail and the frame will be dropped.

If methods are supposed to be blocking, shouldn't `Connect()` actually wait for the peer connection to be connected to return, to ensure that `obs_output_begin_data_capture2()` is called after tracks are open?

Reply:
@paullouisageneau by design. Just so I don’t need to worry about pacing etc…

AFAIK there is no expectation about what the first frame/buffer is